### DISCUSSION TITLE
FabricJS Image Filters vs CSS filter property

### DISCUSSION THREAD

**Discussion by ewajs** [fabricjs/fabric.js#8304] [General]
Hello all,
First of all, thank you for such an amazing Open Source project. I've been working with FabricJS for the past months and have found it to be an amazing library full of useful features.

I'm posting this discussion here at the request of @ShaMan123 who kindly replied a [Stack Overflow question regarding this same issue](https://stackoverflow.com/questions/73391473/fabric-js-filters-vs-css-filters).

As described there, I've been specifically working on a port of an old product for my company in which we do some user provided image positioning/manipulation and we're migrating the current HTML/CSS/jQuery implementation (basically controlling images as DOM elements styled through CSS and animated/controlled via jQuery) to a Canvas based one (using Fabric).

Our aim is to make this transition seamless/invisible for the end user so we implemented our Fabric solution to emulate our current CSS one and we found something that might be interesting to note here, namely that Fabric's Brightness and Contrast implementations seem to be different in nature from the CSS filter one. I managed to solve the issue by defining my own filter based off of fabric's one (I have absolutely no knowledge of shaders but I managed to do it by comparison).

I'll illustrate with an example (will use the same ones from the SO post if you've already read that):

# Original Image
This is the sample, unfiltered image (Brightness = 0 in FabricJS or 1 in CSS).
<img width="379" alt="XbyoD" src="https://user-images.githubusercontent.com/26854120/191136118-ced32d5e-12e3-4c9e-b7ef-d66975ec2c66.png">

# FabricJS
The Brightness filter is additive, [see filter definition](https://github.com/fabricjs/fabric.js/blob/master/src/filters/brightness_filter.class.ts). This has the effect of bringing blacks to white as brightness is increased.

This example has the Brightness set to +33%/0.33, in fabric terms:
```javascript
new fabric.Image.filters.Brightness({ brightness: 0.33 });
```

And the resulting image:
<img width="438" alt="Q6gCc" src="https://user-images.githubusercontent.com/26854120/191136383-30a28e5a-0ab2-4bb6-8283-1627ba1598a3.png">

# CSS `filter` property
The CSS `filter` property accepts a brightness filter (no brightness is a value of 1). However in CSS the brightness filter is multiplicative, [see MDN brightness filter docs](https://developer.mozilla.org/en-US/docs/Web/CSS/filter-function/brightness). This has the effect of making blacks stay black whereas non black parts of the image have their brightness increased (multiplicatively of course). 

This example has the brightness filter property set to +33% (1.33 in CSS filter terms)
```css
img {
      filter: brightness(1.33);
}
```

And the resulting image:
<img width="381" alt="wjkJi" src="https://user-images.githubusercontent.com/26854120/191136739-06d1263e-e11a-4f13-84df-64bb0e2f3484.png">

After doing some digging in order to make both these implementations coherent I eventually found my way to the source code of the Fabric Brightness filter implementation and found that my suspicion of additivieness vs multiplicativeness was correct and just created a multiplicative version of the filter (my implementation might be awful based on my lack of shader experience mentioned before):

# Fabric Multiplicative Brightness Filter
```javascript
// Base implementation taken from https://github.com/fabricjs/fabric.js/blob/master/src/filters/brightness_filter.class.ts
(function(global) {

    'use strict';
  
    var fabric  = global.fabric || (global.fabric = { }),
        filters = fabric.Image.filters,
        createClass = fabric.util.createClass;
  
    /**
     * Multiplicative Brightness filter class
     * @class fabric.Image.filters.MultiplicativeBrightness
     * @memberOf fabric.Image.filters
     * @extends fabric.Image.filters.BaseFilter
     * @example
     * var filter = new fabric.Image.filters.MultiplicativeBrightness({
     *   brightness: 0.05
     * });
     * object.filters.push(filter);
     * object.applyFilters();
     */
    filters.MultiplicativeBrightness = createClass(filters.BaseFilter, /** @lends fabric.Image.filters.MultiplicativeBrightness.prototype */ {
  
      /**
       * Filter type
       * @param {String} type
       * @default
       */
      type: 'Brightness',
  
      /**
       * Fragment source for the brightness program
       */
      fragmentSource: 'precision highp float;\n' +
        'uniform sampler2D uTexture;\n' +
        'uniform float uBrightness;\n' +
        'varying vec2 vTexCoord;\n' +
        'void main() {\n' +
          'vec4 color = texture2D(uTexture, vTexCoord);\n' +
          'color.rgb *= (1.0 + uBrightness);\n' +
          'gl_FragColor = color;\n' +
        '}',
  
      /**
       * Brightness value, from -1 to 1.
       * translated to -255 to 255 for 2d
       * 0.0039215686 is the part of 1 that get translated to 1 in 2d
       * @param {Number} brightness
       * @default
       */
      brightness: 0,
  
      /**
       * Describe the property that is the filter parameter
       * @param {String} m
       * @default
       */
      mainParameter: 'brightness',
  
      /**
      * Apply the Brightness operation to a Uint8ClampedArray representing the pixels of an image.
      *
      * @param {Object} options
      * @param {ImageData} options.imageData The Uint8ClampedArray to be filtered.
      */
      applyTo2d: function(options) {
        if (this.brightness === 0) {
          return;
        }
        var imageData = options.imageData,
            data = imageData.data, i, len = data.length,
            brightness = Math.round(this.brightness * 255);
        for (i = 0; i < len; i += 4) {
          data[i] = data[i] * (1 + brightness);
          data[i + 1] = data[i + 1] * (1 + brightness);
          data[i + 2] = data[i + 2] * (1 + brightness);
        }
      },
  
      /**
       * Return WebGL uniform locations for this filter's shader.
       *
       * @param {WebGLRenderingContext} gl The GL canvas context used to compile this filter's shader.
       * @param {WebGLShaderProgram} program This filter's compiled shader program.
       */
      getUniformLocations: function(gl, program) {
        return {
          uBrightness: gl.getUniformLocation(program, 'uBrightness'),
        };
      },
  
      /**
       * Send data from this filter to its shader program's uniforms.
       *
       * @param {WebGLRenderingContext} gl The GL canvas context used to compile this filter's shader.
       * @param {Object} uniformLocations A map of string uniform names to WebGLUniformLocation objects
       */
      sendUniformData: function(gl, uniformLocations) {
        gl.uniform1f(uniformLocations.uBrightness, this.brightness);
      },
    });
  
    /**
     * Create filter instance from an object representation
     * @static
     * @param {Object} object Object to create an instance from
     * @returns {Promise<fabric.Image.filters.Brightness>}
     */
    fabric.Image.filters.MultiplicativeBrightness.fromObject = fabric.Image.filters.BaseFilter.fromObject;
  
  })(typeof exports !== 'undefined' ? exports : window);
```

I think the same case applies more or less to the Contrast filter but here the math is not so trivial nor did I seem to be able to find a good definition of what CSS is doing. I can read and sorta understand fabric's one but I wouldn't know how to adjust it. For our practical purposes we've found that scaling positive numbers by .45 (in Fabric) and leaving negative numbers unchanged was close enough for us (after unshifting because in CSS 100% contrast is no contrast whereas in fabric that would be 0).  

Maybe there are more people with my use case (trying to port some image processing done via CSS to a Canvas based solution) that would benefit of having a fllter that emulates the CSS filter property handy in order to make migrating these kinds of solution over more easily.

Thanks in advance for all your time and work into this amazing library and hope this is interesting/helpful to the dev team.

**Comment by ShaMan123** [comment]
Great post, thanks @ewajs 
@melchiar I regard you as the filter expert
@asturur

**Comment by melchiar** [comment]
Thanks for the well thought out post @ewajs. Yes, fabric's brightness filter isn't actually very representative of typical brightness filter behavior, I actually use Gamma in place of brightness for my own projects with good results.

**Comment by asturur** [comment]
We can add this better brigthness filter ( maybe we should look around different brightness implementation first ) and deprecate the old one while leaving it available for restore issues.
